File: src/ObjectManager/Object.Tes/FilePacks/Records/345-BOOK.Book.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class BOOKRecord : Record, IHaveEDID, IHaveMODL
    {
        public struct DATAField
        {
            public byte Flags; //: Scroll - (1 is scroll, 0 not)
            public byte Teaches; //: SkillId - (-1 is no skill)
            public int Value;
            public float Weight;
            //
            public int EnchantPts;

            public DATAField(UnityBinaryReader r, int dataSize, GameFormatId format)
            {
                if (format == GameFormatId.TES3)
                {
                    Weight = r.ReadLESingle();
                    Value = r.ReadLEInt32();
                    Flags = (byte)r.ReadLEInt32();
                    Teaches = (byte)r.ReadLEInt32();
                    EnchantPts = r.ReadLEInt32();
                    return;
                }
                Flags = r.ReadByte();
                Teaches = r.ReadByte();
                Value = r.ReadLEInt32();
                Weight = r.ReadLESingle();
                EnchantPts = 0;
            }
        }

        public override string ToString() => $"BOOK: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model (optional)
        public STRVField FULL; // Item Name
        public DATAField DATA; // Book Data
        public STRVField DESC; // Book Text
        public FILEField ICON; // Inventory Icon (optional)
        public FMIDField<SCPTRecord> SCRI; // Script Name (optional)
        public FMIDField<ENCHRecord> ENAM; // Enchantment FormId (optional)
        // TES4
        public IN16Field? ANAM; // Enchantment points (optional)

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "FULL":
                case "FNAM": FULL = new STRVField(r, dataSize); return true;
                case "DATA":
                case "BKDT": DATA = new DATAField(r, dataSize, format); return true;
                case "ICON":
                case "ITEX": ICON = new FILEField(r, dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                case "DESC":
                case "TEXT": DESC = new STRVField(r, dataSize); return true;
                case "ENAM": ENAM = new FMIDField<ENCHRecord>(r, dataSize); return true;
                case "ANAM": ANAM = new IN16Field(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-BOOK.Book.swift
//
//  BOOKRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class BOOKRecord: Record, IHaveEDID, IHaveMODL {
    public struct DATAField {
        public let flags: UInt8 //: Scroll - (1 is scroll, 0 not)
        public let teaches: UInt8 //: SkillId - (-1 is no skill)
        public let value: Int32
        public let weight: Float
        //
        public let enchantPts: Int32

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            guard format != .TES3 else {
                weight = r.readLESingle()
                value = r.readLEInt32()
                flags = UInt8(r.readLEInt32())
                teaches = UInt8(checkMax: r.readLEInt32())
                enchantPts = r.readLEInt32()
                return
            }
            flags = r.readByte()
            teaches = r.readByte()
            value = r.readLEInt32()
            weight = r.readLESingle()
            enchantPts = 0
        }
    }

    public override var description: String { return "BOOK: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var MODL: MODLGroup? = nil // Model (optional)
    public var FULL: STRVField! // Item Name
    public var DATA: DATAField! // Book Data
    public var DESC: STRVField! // Book Text
    public var ICON: FILEField? = nil // Inventory Icon (optional)
    public var SCRI: FMIDField<SCPTRecord>? = nil // Script Name (optional)
    public var ENAM: FMIDField<ENCHRecord>? = nil // Enchantment FormId (optional)
    // TES4
    public var ANAM: IN16Field? = nil // Enchantment points (optional)

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL?.MODBField(r, dataSize)
        case "MODT": MODL?.MODTField(r, dataSize)
        case "FULL",
             "FNAM": FULL = r.readSTRV(dataSize)
        case "DATA",
             "BKDT": DATA = DATAField(r, dataSize, format)
        case "ICON",
             "ITEX": ICON = r.readSTRV(dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        case "DESC",
             "TEXT": DESC = r.readSTRV(dataSize)
        case "ENAM": ENAM = FMIDField<ENCHRecord>(r, dataSize)
        case "ANAM": ANAM = r.readT(dataSize)
        default: return false
        }
        return true
    }
}
